### PR TITLE
feat:  export persist_period_sec option and background_threads

### LINKED ISSUE #447: export stats_persist_period_sec option and rocksdb_env_set_background_threads
These are needed if user want to disable all threads in rocksdb. Only set `set_max_background_jobs`, `set_max_background_compactions`, `set_max_background_flushes` to 0 will still leave 4 additional threads. By this in C++ rocksdb it's possible to disable all threads by (tested in lldb):
```c++
  options.max_background_jobs = 0; // already have in rust-rocksdb, master
  options.max_background_compactions = 0;  // already have in rust-rocksdb 0.14
  options.max_background_flushes = 0;  // already have in rust-rocksdb 0.14
  options.stats_dump_period_sec = 0;  // already have in rust-rocksdb 0.14
  options.stats_persist_period_sec = 0; // **need export**
  DB* db = nullptr;
  Status s = DB::Open(options, kDBPath, &db);
  Env::Default()->SetBackgroundThreads(0, Env::Priority::BOTTOM); // **need export**
  Env::Default()->SetBackgroundThreads(0, Env::Priority::LOW);
  Env::Default()->SetBackgroundThreads(0, Env::Priority::HIGH);
```

### PR DESCRIPTION
**Pull request by developerfred**
exported functions:
- [x] `set_low_priority_background_threads`
- [x] `set_bottom_priority_background_threads`
- [x]  `env_default`  https://github.com/facebook/rocksdb/pull/7207
- [x] `set_stats_persist_period_sec`

close #447

Export PR on RocksDB
https://github.com/facebook/rocksdb/pull/7168

### CODE REVIEW

**Comment by aleksuss** [comment]
Thanks for the contributing. Overall looks good. But could you add some tests ?

**Comment by developerfred** [comment]
@aleksuss Yes! I'll add now. thanks for the quick feedback

File: src/db_options.rs [hunk 1]
```diff
@@ -136,18 +136,18 @@ impl Env {
             ffi::rocksdb_env_set_high_priority_background_threads(self.inner, n);
         }
     }
-
-    /// Wait for all threads started by StartThread to terminate.
-    pub fn join_all_threads(&mut self) {
+    
+    /// Sets the size of the bottom priority thread pool 
+    /// 
+    pub fn set_bottom_priority_background_threads(&mut self, n: c_int) {
```
Comment (aleksuss) [inline src/db_options.rs:142]: You call `ffi::rocksdb_env_set_high_priority_background_threads` in method which name includes `*bottom* ? 
If you look into C code than you'll see that this function has such code inside:
```cpp
void rocksdb_env_set_high_priority_background_threads(rocksdb_env_t* env, int n) {
  env->rep->SetBackgroundThreads(n, Env::HIGH);
}
```

File: src/db_options.rs [hunk 2]
```diff
@@ -2188,6 +2196,24 @@ impl Options {
...
+    ///
+    /// ```
+    /// use rocksdb::Options;
+    ///
+    /// let mut opts = Options::default();
+    /// opts.set_stats_dump_period_sec(300);
+    /// ```
+    pub fn set_stats_persist_period_sec(&mut self, period: c_uint) {
+        unsafe {
+            ffi:rocksdb_options_set_stats_persist_period_sec(self.inner, period);
```
Comment (aleksuss) [inline src/db_options.rs:2213]:
```suggestion
            ffi::rocksdb_options_set_stats_persist_period_sec(self.inner, period);
```

File: src/db_options.rs (hunk 1 above)
Comment (developerfred) [reply src/db_options.rs]: Thanks, I removed

File: src/db_options.rs [hunk 3]
```diff
@@ -2188,6 +2189,24 @@ impl Options {
...
+    ///
+    /// ```
+    /// use rocksdb::Options;
+    ///
+    /// let mut opts = Options::default();
+    /// opts.set_stats_persist_period_sec(0);
+    /// ```
+    pub fn set_stats_persist_period_sec(&mut self, period: c_uint) {
+        unsafe {
+            ffi::rocksdb_options_set_stats_persist_period_sec(self.inner, period);
```
Comment (DarkEld3r) [inline src/db_options.rs:2206]: As far as I can see, the `rocksdb_options_set_stats_persist_period_sec` function isn't present in the rocksdb C-api. Looks like it should be exported.

Comment (developerfred) [reply src/db_options.rs:2206]: @DarkEld3r I will export

Comment (developerfred) [reply src/db_options.rs:2205]: @DarkEld3r https://github.com/facebook/rocksdb/pull/7168

**Comment by ailisp** [comment]
@developerfred expose `set_stats_persist_period_sec` looks good, but disable background threads isn't done. `Env::Default()->SetBackgroundThreads(0, Env::Priority::BOTTOM); // **need export**` is a different function than `max_background_jobs`. Which you can tested by:
1. launch program with lldb, place a break point before:
```
             assert_eq!(k.as_ref(), format!("{:0>4}", expected).as_bytes());
```
2. run `thread list`

Without `Env::Default()->SetBackgroundThreads` you'll see 3+ threads. (set_stats_persist_period_sec will turn off one thread, but there's still two other background threads). Expected only 1 thread with `Env::Default()->SetBackgroundThreads`, for example, run lldb with this: https://github.com/nearprotocol/rocksdb/blob/no-thread-test/examples/compact_files_example.cc

**Comment by developerfred** [comment]
@ailisp I will open a new pull request to export the `Env::Default()->SetBackgroundThreads(0, Env::Priority::BOTTOM);`

**Comment by developerfred** [comment]
@ailisp I exported how rocksdb_env_set_background_threads_disable

**Comment by ailisp** [comment]
> @ailisp I exported how `rocksdb_env_set_background_threads_disable`

It might be good idea to follow their naming: https://github.com/facebook/rocksdb/pull/7191/files#diff-da9c25aee06e97392424d9a63a849f10R3834

**Comment by developerfred** [comment]
@ailisp done 👍🏾  - `rocksdb_env_set_bottom_priority_background_threads` and `rocksdb_env_set_low_priority_background_threads`

**Comment by developerfred** [comment]
@ailisp  Bump done on 4de43cccf23a6947bd4ce443f3627a549beb2f49

**Comment by ailisp** [comment]
> @ailisp Bump done
2c3d7d7

:+1 Can you export them in rust-rocksdb and call them in test? I think that'll be the last step we need

**Comment by developerfred** [comment]
> > @ailisp Bump done
> > [2c3d7d7](https://github.com/rust-rocksdb/rust-rocksdb/commit/2c3d7d7dc59f6acbd8e944b74a7a7dc2baed5bd4)
> 
> :+1 Can you export them in rust-rocksdb and call them in test? I think that'll be the last step we need

done :white_check_mark:

**Comment by ailisp** [comment]
> @ailisp Perhaps you can depend on this pull request branch until a new version is released?
Yeah, i can depends on his fork until new release.

@developerfred I tested your commit, it looks good, however due to rocksdb's behavior, SetMaxBackgroundThreads only have effect after open database:
This have effects:
https://github.com/nearprotocol/rocksdb/blob/no-thread-test/examples/compact_files_example.cc#L149
But set it before open database have no effect (commented out above).

To see whether it has effect (only 1 thread in lldb shows it have effect)
image

So i made this repo, basically copied your test code but make it a separate binary to avoid in case `cargo test` create any thread.
https://github.com/ailisp/test-disable-thread
Then
```
cargo build
target/debug/test-disable-thread
ps aux | grep test-disable-thread
lldb
lldb> attach -p <pid>
lldb> thread list
```

but there's still show three thread:


This implies Env::default() is a different thing than c++ rocksdb's Env::Default() obj (I check and it looks so in source code). So we need to export it as well :(

**Comment by developerfred** [comment]
@ailisp I have a local limitation to do the tests, I don't have enough disk space. I'm trying to use gitpod but without llvm success. My next step is to delete some software from my mac.

Do you have a sense of the space needed?

**Comment by ailisp** [comment]
@developerfred My bad in test, your PRs fully works, it just takes some time to wait rocksdb destroy other threads (wait some seconds and lldb attach there's only one thread). I'm very sorry of that. So no need to export Env::Default (ust-rocksdb/rust-rocksdb#448)

**Comment by developerfred** [comment]
@ailisp ok! No problem!

**Comment by ailisp** [comment]
@developerfred We've successfully use your fork on our repo, Thank you! looking forward to next rust-rocksdb release, to upgrade your changes in c++ rocksdb and land this PR

**Comment by ailisp** [comment]
@stanislav-tkach Facebook has released rocksdb containing @developerfred 's change and it's been bumped in rust-rocksdb, let's merge this PR? (@developerfred Would you mind resolve conflict? I cannot push to your branch, if you don't have time I can fork from your fork, resolve conflict and make a new PR, then close this one)

**Comment by developerfred** [comment]
> @stanislav-tkach Facebook has released rocksdb containing @developerfred 's change and it's been bumped in rust-rocksdb, let's merge this PR? (@developerfred Would you mind resolve conflict? I cannot push to your branch, if you don't have time I can fork from your fork, resolve conflict and make a new PR, then close this one)

I solve this now, thanks for information and ping.

**Comment by aleksuss** [comment]
@developerfred apply please suggestions by `fmt` and `clippy`.

**Comment by developerfred** [comment]
ok @aleksuss

**Review by aleksuss — APPROVED** [review]

**Comment by ailisp** [comment]
Thank you @developerfred and @aleksuss !

**Pull request merged**